### PR TITLE
Sf2 megaverse

### PR DESCRIPTION
**Pull request by wmFrank**
- integrate megaverse into sf2
- report signal_queue size in heartbeat

### CODE REVIEW

**Comment by alex-petrenko** [comment]
Looks like there are merge conflicts with your previous commit @wmFrank 
Can you please resolve this?

File: sample_factory/algo/runners/runner.py [hunk 1]
```diff
@@ -505,9 +506,17 @@ def _setup_component_heartbeat(self, component: HeartbeatStoppableEventLoopObjec
             self.heartbeat_dict[component_type] = {}
         type_dict = self.heartbeat_dict[component_type]
         type_dict[component.object_id] = time.time()
+
+        """
```
Comment (alex-petrenko) [inline sample_factory/algo/runners/runner.py:510]: This is not how one usually uses docstrings. Docstrings are intended for functions and classes, if you want to add a comment in the middle of the function, just use regular comment starting with `#`

File: sample_factory/algo/runners/runner.py [hunk 2]
```diff
@@ -505,9 +506,17 @@ def _setup_component_heartbeat(self, component: HeartbeatStoppableEventLoopObjec
...
+        setup up queue_size report with heartbeat, grouped by event_loop_process_id
+        """
+        process_id = 0 if component.event_loop.process is None else component.event_loop.process.ident
+        if process_id not in self.queue_size_dict:
+            self.queue_size_dict[process_id] = 0
+
         component.heartbeat.connect(self._receive_heartbeat)
 
-    def _receive_heartbeat(self, component_type: type, component_id: str):
+    def _receive_heartbeat(self, component_type: type, component_id: str, process_id, qsize):
```
Comment (alex-petrenko) [inline sample_factory/algo/runners/runner.py:519]: type hints would be nice! :)

File: sample_factory/algo/runners/runner.py [hunk 3]
```diff
@@ -541,6 +551,10 @@ def _check_heartbeat(self):
             log.error("Stopping training from lack of heartbeats from " + ", ".join(type_list))
             self._stop_training()
 
+        log.debug("Checking Queue Size")
```
Comment (alex-petrenko) [inline sample_factory/algo/runners/runner.py:554]: I think this would spam the logs too much.
Let's:
- remove "Checking queue size" log
- for the other log let's print using debug_log_every_n function with n=1000
- is process id a number here? It should be a unique name of the process or event loop. You can get it from EventLoop or EventLoopProcess class

File: sample_factory/algo/sampling/rollout_worker.py [hunk 4]
```diff
@@ -36,6 +43,7 @@ def init_rollout_worker_process(sf_context: SampleFactoryContext, worker: Rollou
     os_signal.signal(os_signal.SIGINT, os_signal.SIG_IGN)
 
     cfg = worker.cfg
+    init_file_logger(experiment_dir(cfg))
```
Comment (alex-petrenko) [inline sample_factory/algo/sampling/rollout_worker.py:46]: These are probably merge conflicts

File: sample_factory/algo/utils/heartbeat.py [hunk 5]
```diff
@@ -18,7 +18,9 @@ def stop(self):
         ...
 
     def _report_heartbeat(self):
-        self.heartbeat.emit(type(self), self.object_id)
+        process_id = 0 if self.event_loop.process is None else self.event_loop.process.ident
```
Comment (alex-petrenko) [inline sample_factory/algo/utils/heartbeat.py:21]: Let's use process name instead of ID here. Or both name and ID.

File: sample_factory/cfg/cfg.py [hunk 6]
```diff
@@ -553,6 +553,13 @@ def add_default_env_args(p: ArgumentParser):
         type=str2bool,
         help="Set to true if environment expects actions on GPU (i.e. as a GPU-side PyTorch tensor)",
     )
+    p.add_argument(
+        "--env_gpu_observations",
+        default=True,
+        type=str2bool,
+        help="Set to true if environment expects data on GPU, If this is true we expect data to be on the device determined by actor_worker_gpus, otherwise it will be on CPU",
```
Comment (alex-petrenko) [inline sample_factory/cfg/cfg.py:560]: Okay this is good.

**Review by alex-petrenko — COMMENTED** [review]
Please address my comments.

Also we should:
1) Do a documentation page that points to Megaverse installation instructions
2) Run some experiments on TowerBuilding with 3-5 seeds and attach a WandB report to docs
3) Ideally upload to HuggingFace

File: sample_factory/algo/runners/runner.py [hunk 7]
```diff
@@ -559,6 +568,9 @@ def _check_heartbeat(self):
             log.error(f"Stopping training from lack of heartbeats from {', '.join(type_list)}")
             self._stop_training()
 
+        for p_name, qsize in self.queue_size_dict.items():
```
Comment (alex-petrenko) [inline sample_factory/algo/runners/runner.py:571]: Maybe say here that the process has <N> messages in the outgoing signal queue.

And let's only report this if qsize is bigger than some threshold, like 5.

File: docs/environment-integrations/megaverse.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+Or you could run experiments on slurm:
+
+```
+python -m sample_factory.runner.run --runner=slurm --slurm_workdir=./slurm_megaverse --experiment_suffix=slurm --slurm_gpus_per_job=1 --slurm_cpus_per_gpu=16 --slurm_sbatch_template=./sample_factory/runner/slurm/sbatch_template.sh --pause_between=1 --slurm_print_only=False --run=sf_examples.megaverse_examples.experiments.megaverse_envs
+```
+
+
+### Results
+
+#### Reports
```
Comment (alex-petrenko) [inline docs/environment-integrations/megaverse.md:38]: Why we only have the report for 4 agents?
Can we put the report for 1 agent here instead?

File: sf_examples/megaverse_examples/experiments/megaverse_envs.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+from sample_factory.runner.run_description import Experiment, ParamGrid, RunDescription
+
+_params = ParamGrid(
+    [
+        ("seed", [00, 11, 22, 33]),
+        (
+            "env",
+            [
+                "megaverse_TowerBuilding",
```
Comment (alex-petrenko) [inline sf_examples/megaverse_examples/experiments/megaverse_envs.py:9]: Since the experiment is called "megaverse_envs", can we add all other envs here in the "env" list?
But maybe comment them out. This way it would be easy for ppl to understand how to run with other envs (just uncomment them).

File: sf_examples/megaverse_examples/experiments/megaverse_envs.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+            ],
+        ),
+    ]
+)
+
+_experiments = [
+    Experiment(
+        "megaverse_envs",
+        "python -m sf_examples.megaverse_examples.train_megaverse --train_for_seconds=360000000 --train_for_env_steps=2000000000 --algo=APPO --gamma=0.997 --use_rnn=True --rnn_num_layers=2 --num_workers=10 --num_envs_per_worker=2 --num_epochs=1 --rollout=32 --recurrence=32 --batch_size=4096 --actor_worker_gpus 0 --env_gpu_observations=False --num_policies=1 --with_pbt=False --max_grad_norm=0.0 --exploration_loss=symmetric_kl --exploration_loss_coeff=0.001 --megaverse_num_simulation_threads=1 --megaverse_num_envs_per_instance=30 --megaverse_num_agents_per_env=4 --megaverse_use_vulkan=True --policy_workers_per_policy=2 --reward_clip=30 --with_wandb=True --wandb_project=megaverse-benchmark --wandb_group=megaverse --wandb_tags sf2 run5",
+        _params.generate_params(randomize=False),
```
Comment (alex-petrenko) [inline sf_examples/megaverse_examples/experiments/megaverse_envs.py:19]: Change `megaverse_num_agents_per_env` to 1 here. And maybe increase `megaverse_num_envs_per_instance` to like 80 so the total number of agents is similar (or whatever value you used for the single agent experiment)

File: sf_examples/megaverse_examples/megaverse/megaverse_params.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    parser.set_defaults(
+        encoder_type="conv",
+        encoder_subtype="convnet_simple",
+        hidden_size=512,
+        obs_subtract_mean=0.0,
+        obs_scale=255.0,
+        actor_worker_gpus=[0],
+        env_gpu_observations=False,
+        exploration_loss="symmetric_kl",
+        exploration_loss_coeff=0.001,
```
Comment (alex-petrenko) [inline sf_examples/megaverse_examples/megaverse/megaverse_params.py:15]: Let's enable normalizations by default:
`--normalize_input=True --normalize_returns=True`

File: sf_examples/megaverse_examples/megaverse/megaverse_utils.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+from sample_factory.utils.utils import log
+
+
+class MegaverseSpec:
+    def __init__(self, name):
+        self.name = name
+
+
+MEGAVERSE_ENVS = [
+    MegaverseSpec("megaverse_TowerBuilding"),
```
Comment (alex-petrenko) [inline sf_examples/megaverse_examples/megaverse/megaverse_utils.py:14]: Why not also have all the other specs?
Can you just copy them from Megaverse repo?

Comment (alex-petrenko) [reply sf_examples/megaverse_examples/megaverse/megaverse_utils.py:14]: Also, is there any reason to use `megaverse_` prefix for everything? Can we just do `TowerBuilding`?

File: sf_examples/megaverse_examples/megaverse/megaverse_utils.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        obs, rewards, dones, infos = self.env.step(action)
+
+        for i, info in enumerate(infos):
+            self.episode_rewards[i] += rewards[i]
+
+            if dones[i]:
+                if "episode_extra_stats" not in info:
+                    info["episode_extra_stats"] = dict()
+                extra_stats = info["episode_extra_stats"]
+                extra_stats[f"z_{self.env.unwrapped.scenario_name.casefold()}_true_reward"] = info["true_reward"]
```
Comment (alex-petrenko) [inline sf_examples/megaverse_examples/megaverse/megaverse_utils.py:59]: `true_reward` here should be renamed to `true_objective`, because this is what we use in SF2 now.

Can you just do info["true_objective"] = info["true_reward"] so we have both?
Also rename summaries to `..._true_objective`

File: sf_examples/megaverse_examples/megaverse/megaverse_utils.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+                self.episode_rewards[i] = 0
+
+                if self.increase_team_spirit:
+                    rew_shaping = self.get_current_reward_shaping(i)
+                    rew_shaping["teamSpirit"] = min(approx_total_training_steps / self.max_team_spirit_steps, 1.0)
+                    self.set_reward_shaping(rew_shaping, i)
+                    extra_stats["teamSpirit"] = rew_shaping["teamSpirit"]
+
+        terminated = dones
+        truncated = dones
```
Comment (alex-petrenko) [inline sf_examples/megaverse_examples/megaverse/megaverse_utils.py:74]: This is not correct.
Megaverse currently does not differentiate between envs terminated by timer and by reaching terminal states.
So it is wrong to assume that all episodes finish by truncation (timeout). It should be `truncated=False`

It would be nice to get this information from the env, but definitely not right now :)

**Review by alex-petrenko — CHANGES_REQUESTED** [review]
Very good work!

We have a few points to address. See my comments!

**Comment by alex-petrenko** [comment]
Consider also removing megaverse from CI coverage analysis because we can't run tests for this on github actions even if we had them.

**Review by alex-petrenko — APPROVED** [review]
